File: api_client.py
import requests
import json
import time
from typing import List, Dict, Any
import logging

class APIClient:
    def __init__(self, api_url: str):
        self.api_url = api_url
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'ZKTeco-Service/2.0'
        })
        # Désactiver la vérification SSL pour les tests
        self.session.verify = False
        import urllib3
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
        
        self.logger = logging.getLogger(__name__)
    
    def send_attendance(self, attendance_data: List[Dict[str, Any]], max_retries: int = 3) -> bool:
        """Envoyer les pointages à l'API avec système de retry"""
        if not attendance_data:
            self.logger.info("📭 Aucune donnée à envoyer")
            return True
        
        for attempt in range(max_retries):
            try:
                self.logger.info(f"📤 Tentative {attempt + 1}/{max_retries} - Envoi de {len(attendance_data)} pointages")
                
                response = self.session.post(
                    self.api_url,
                    json=attendance_data,
                    timeout=30,
                    verify=False
                )
                
                if response.status_code in [200, 201]:
                    try:
                        response_data = response.json()
                        self.logger.info(f"✅ Réponse API: {response_data.get('message', 'Succès')}")
                        self.logger.info(f"📊 {response_data.get('saved_count', 0)} sauvegardés, {response_data.get('duplicates_skipped', 0)} doublons ignorés")
                        return True
                    except ValueError:
                        self.logger.info("✅ Pointages envoyés avec succès")
                        return True
                else:
                    self.logger.warning(f"⚠️  Réponse API {response.status_code}: {response.text}")
                    
                    if attempt < max_retries - 1:
                        wait_time = (attempt + 1) * 10
                        self.logger.info(f"⏳ Nouvelle tentative dans {wait_time}s...")
                        time.sleep(wait_time)
                    
            except requests.exceptions.ConnectionError as e:
                self.logger.error(f"🔌 Erreur connexion API: {e}")
            except requests.exceptions.Timeout as e:
                self.logger.error(f"⏰ Timeout API: {e}")
            except Exception as e:
                self.logger.error(f"💥 Erreur inattendue: {e}")
            
            if attempt < max_retries - 1:
                time.sleep((attempt + 1) * 10)
        
        self.logger.error(f"🔴 Échec après {max_retries} tentatives")
        return False
    
    def test_connection(self) -> bool:
        """Tester la connexion à l'API de manière plus robuste"""
        try:
            self.logger.info(f"🧪 Test de connexion API: {self.api_url}")
            
            # Méthode 1: Essayer une requête OPTIONS (plus safe)
            try:
                response = self.session.options(self.api_url, timeout=10, verify=False)
                if response.status_code in [200, 204, 405]:  # 405 Method Not Allowed est acceptable
                    self.logger.info("✅ Test OPTIONS réussi")
                    return True
            except:
                pass
            
            # Méthode 2: Essayer une requête HEAD
            try:
                response = self.session.head(self.api_url, timeout=10, verify=False)
                if response.status_code != 404:
                    self.logger.info("✅ Test HEAD réussi")
                    return True
            except:
                pass
            
            # Méthode 3: Essayer l'endpoint racine sans /pointages
            try:
                base_url = self.api_url.rsplit('/pointages', 1)[0]
                if base_url != self.api_url:  # Vérifier qu'on a bien enlevé /pointages
                    response = self.session.get(base_url, timeout=10, verify=False)
                    if response.status_code != 404:
                        self.logger.info("✅ Test endpoint racine réussi")
                        return True
            except:
                pass
            
            # Méthode 4: Essayer avec des données vides
            try:
                response = self.session.post(
                    self.api_url,
                    json=[],
                    timeout=10,
                    verify=False
                )
                # Même une erreur 400 signifie que l'endpoint existe
                if response.status_code != 404:
                    self.logger.info("✅ Test avec données vides réussi")
                    return True
            except requests.exceptions.ConnectionError:
                self.logger.error("🔌 Impossible de se connecter à l'API")
                return False
            except Exception as e:
                # D'autres erreurs peuvent signifier que l'endpoint existe mais refuse les données
                if "404" not in str(e):
                    self.logger.info("✅ Endpoint accessible (erreur autre que 404)")
                    return True
            
            self.logger.error("❌ Aucune méthode de test n'a fonctionné")
            return False
            
        except Exception as e:
            self.logger.error(f"❌ Test connexion API échoué: {e}")
            return False

File: ZKTecoService/src/com/zkteco/a/ApiClient.java
package com.zkteco;

import java.io.*;
import java.net.*;
import java.util.*;
import java.util.logging.*;
import javax.net.ssl.HttpsURLConnection;
public class ApiClient {
    private static final Logger logger = Logger.getLogger(ApiClient.class.getName());
    private String apiUrl;
    
    public ApiClient(String apiUrl) {
        this.apiUrl = apiUrl;
    }
    
    public boolean sendAttendance(List<Map<String, Object>> attendanceData) {
        return sendAttendance(attendanceData, 3);
    }
    
    public boolean sendAttendance(List<Map<String, Object>> attendanceData, int maxRetries) {
        if (attendanceData == null || attendanceData.isEmpty()) {
            logger.info("📭 Aucune donnée à envoyer");
            return true;
        }
        
        for (int attempt = 0; attempt < maxRetries; attempt++) {
            try {
                logger.info("📤 Tentative " + (attempt + 1) + "/" + maxRetries + 
                           " - Envoi de " + attendanceData.size() + " pointages");
                
                URL url = new URL(apiUrl);
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setRequestProperty("User-Agent", "ZKTeco-Service/2.0");
                connection.setConnectTimeout(30000);
                connection.setReadTimeout(30000);
                connection.setDoOutput(true);
                
                // Désactiver la vérification SSL
                if (connection instanceof HttpsURLConnection) {
                    ((HttpsURLConnection) connection).setSSLSocketFactory(
                        getTrustAllSocketFactory());
                }
                
                // Convertir les données en JSON
                String jsonData = convertToJson(attendanceData);
                
                try (OutputStream os = connection.getOutputStream()) {
                    byte[] input = jsonData.getBytes("utf-8");
                    os.write(input, 0, input.length);
                }
                
                int responseCode = connection.getResponseCode();
                
                if (responseCode == 200 || responseCode == 201) {
                    String response = readResponse(connection);
                    logger.info("✅ Réponse API: " + response);
                    return true;
                } else {
                    logger.warning("⚠️ Réponse API " + responseCode + ": " + readError(connection));
                    
                    if (attempt < maxRetries - 1) {
                        int waitTime = (attempt + 1) * 10;
                        logger.info("⏳ Nouvelle tentative dans " + waitTime + "s...");
                        Thread.sleep(waitTime * 1000);
                    }
                }
                
            } catch (SocketTimeoutException e) {
                logger.severe("⏰ Timeout API: " + e.getMessage());
            } catch (ConnectException e) {
                logger.severe("🔌 Erreur connexion API: " + e.getMessage());
            } catch (Exception e) {
                logger.severe("💥 Erreur inattendue: " + e.getMessage());
            }
            
            if (attempt < maxRetries - 1) {
                try {
                    Thread.sleep((attempt + 1) * 10000);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        }
        
        logger.severe("🔴 Échec après " + maxRetries + " tentatives");
        return false;
    }
    
    public boolean testConnection() {
        try {
            logger.info("🧪 Test de connexion API: " + apiUrl);
            
            // Méthode 1: OPTIONS
            try {
                URL url = new URL(apiUrl);
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("OPTIONS");
                connection.setConnectTimeout(10000);
                connection.setReadTimeout(10000);
                
                if (connection instanceof HttpsURLConnection) {
                    ((HttpsURLConnection) connection).setSSLSocketFactory(
                        getTrustAllSocketFactory());
                }
                
                int responseCode = connection.getResponseCode();
                if (responseCode == 200 || responseCode == 204 || responseCode == 405) {
                    logger.info("✅ Test OPTIONS réussi");
                    return true;
                }
            } catch (Exception e) {
                // Continuer avec d'autres méthodes
            }
            
            // Méthode 2: HEAD
            try {
                URL url = new URL(apiUrl);
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("HEAD");
                connection.setConnectTimeout(10000);
                connection.setReadTimeout(10000);
                
                if (connection instanceof HttpsURLConnection) {
                    ((HttpsURLConnection) connection).setSSLSocketFactory(
                        getTrustAllSocketFactory());
                }
                
                int responseCode = connection.getResponseCode();
                if (responseCode != 404) {
                    logger.info("✅ Test HEAD réussi");
                    return true;
                }
            } catch (Exception e) {
                // Continuer avec d'autres méthodes
            }
            
            // Méthode 3: POST avec données vides
            try {
                URL url = new URL(apiUrl);
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("POST");
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setConnectTimeout(10000);
                connection.setReadTimeout(10000);
                connection.setDoOutput(true);
                
                if (connection instanceof HttpsURLConnection) {
                    ((HttpsURLConnection) connection).setSSLSocketFactory(
                        getTrustAllSocketFactory());
                }
                
                // Envoyer tableau vide
                String emptyJson = "[]";
                try (OutputStream os = connection.getOutputStream()) {
                    byte[] input = emptyJson.getBytes("utf-8");
                    os.write(input, 0, input.length);
                }
                
                int responseCode = connection.getResponseCode();
                if (responseCode != 404) {
                    logger.info("✅ Test avec données vides réussi");
                    return true;
                }
            } catch (ConnectException e) {
                logger.severe("🔌 Impossible de se connecter à l'API");
                return false;
            } catch (Exception e) {
                if (!e.getMessage().contains("404")) {
                    logger.info("✅ Endpoint accessible (erreur autre que 404)");
                    return true;
                }
            }
            
            logger.severe("❌ Aucune méthode de test n'a fonctionné");
            return false;
            
        } catch (Exception e) {
            logger.severe("❌ Test connexion API échoué: " + e.getMessage());
            return false;
        }
    }
    
    private String convertToJson(List<Map<String, Object>> data) {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = 0; i < data.size(); i++) {
            Map<String, Object> item = data.get(i);
            sb.append("{");
            int count = 0;
            for (Map.Entry<String, Object> entry : item.entrySet()) {
                sb.append("\"").append(entry.getKey()).append("\":");
                if (entry.getValue() instanceof String) {
                    sb.append("\"").append(entry.getValue()).append("\"");
                } else {
                    sb.append(entry.getValue());
                }
                if (++count < item.size()) {
                    sb.append(",");
                }
            }
            sb.append("}");
            if (i < data.size() - 1) {
                sb.append(",");
            }
        }
        sb.append("]");
        return sb.toString();
    }
    
    private String readResponse(HttpURLConnection connection) {
        try {
            BufferedReader in = new BufferedReader(
                new InputStreamReader(connection.getInputStream()));
            String inputLine;
            StringBuilder content = new StringBuilder();
            while ((inputLine = in.readLine()) != null) {
                content.append(inputLine);
            }
            in.close();
            return content.toString();
        } catch (Exception e) {
            return "Erreur lecture réponse: " + e.getMessage();
        }
    }
    
    private String readError(HttpURLConnection connection) {
        try {
            BufferedReader in = new BufferedReader(
                new InputStreamReader(connection.getErrorStream()));
            String inputLine;
            StringBuilder content = new StringBuilder();
            while ((inputLine = in.readLine()) != null) {
                content.append(inputLine);
            }
            in.close();
            return content.toString();
        } catch (Exception e) {
            return "Erreur lecture erreur: " + e.getMessage();
        }
    }
    
    private javax.net.ssl.SSLSocketFactory getTrustAllSocketFactory() {
        try {
            javax.net.ssl.TrustManager[] trustAllCerts = new javax.net.ssl.TrustManager[]{
                new javax.net.ssl.X509TrustManager() {
                    public java.security.cert.X509Certificate[] getAcceptedIssuers() {
                        return null;
                    }
                    public void checkClientTrusted(
                        java.security.cert.X509Certificate[] certs, String authType) {
                    }
                    public void checkServerTrusted(
                        java.security.cert.X509Certificate[] certs, String authType) {
                    }
                }
            };
            
            javax.net.ssl.SSLContext sc = javax.net.ssl.SSLContext.getInstance("SSL");
            sc.init(null, trustAllCerts, new java.security.SecureRandom());
            return sc.getSocketFactory();
        } catch (Exception e) {
            return null;
        }
    }
}